File: authorization/src/main/kotlin/com/example/authorization/config/SecurityConfig.java
package com.example.authorization.config;

import com.example.authorization.filter.*;
import org.springframework.context.annotation.*;
import org.springframework.security.config.*;
import org.springframework.security.config.annotation.web.builders.*;
import org.springframework.security.config.http.*;
import org.springframework.security.crypto.bcrypt.*;
import org.springframework.security.crypto.password.*;
import org.springframework.security.web.*;
import org.springframework.security.web.authentication.www.*;
import org.springframework.web.cors.*;

import java.time.*;
import java.util.*;

import static com.example.authorization.constants.SecurityConstants.*;

@Configuration
public class SecurityConfig {

    @Bean
    SecurityFilterChain defaultSecurityFilterChain(HttpSecurity http) throws Exception {
        http
                .csrf().disable()
                .sessionManagement()
                .sessionCreationPolicy(
                        SessionCreationPolicy.STATELESS
                )
                .and()
                .cors()
                .configurationSource(
                        corsConfigurationSource()
                )
                .and()
                .authorizeHttpRequests((auth) -> auth
                        .requestMatchers("/api/user-info").authenticated()
                        .requestMatchers("/api/register", "/api/login").permitAll()
                )
                .httpBasic(Customizer.withDefaults())
                .addFilterBefore(
                        new JWTTokenValidatorFilter(JWT_HEADER, JWT_KEY), BasicAuthenticationFilter.class
                );

        return http.build();
    }

    @Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration config = new CorsConfiguration();
        config.setAllowedOrigins(Collections.singletonList("*"));
        config.setAllowedMethods(Collections.singletonList("*"));
        config.setAllowCredentials(true);
        config.setAllowedHeaders(Collections.singletonList("*"));
        config.setExposedHeaders(List.of("Authorization"));
        config.setMaxAge(Duration.ofHours(1));

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", config);
        return source;
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }
}

File: authorization/src/main/kotlin/com/example/authorization/config/SecurityConfig.kt
package com.example.authorization.config

import com.example.authorization.constants.SecurityConstants.JWT_HEADER
import com.example.authorization.constants.SecurityConstants.JWT_KEY
import com.example.authorization.filter.JWTTokenValidatorFilter
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter
import org.springframework.web.cors.*

@Configuration
@EnableWebSecurity
class SecurityConfig {

    @Bean
    fun defaultSecurityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http.csrf().disable()
            .sessionManagement()
            .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
            .and()
            .cors()
            .configurationSource(corsConfigurationSource())
            .and()
            .authorizeHttpRequests { auth ->
                auth.requestMatchers("/api/user-info").authenticated()
                    .requestMatchers("/api/register", "/api/login").permitAll()
            }
            .httpBasic(Customizer.withDefaults())
            .addFilterBefore(
                JWTTokenValidatorFilter(JWT_HEADER, JWT_KEY),
                BasicAuthenticationFilter::class.java
            )

        return http.build()
    }

    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val config = CorsConfiguration()
        config.allowedOrigins = listOf("*")
        config.allowedMethods = listOf("*")
        config.allowCredentials = true
        config.allowedHeaders = listOf("*")
        config.exposedHeaders = listOf("Authorization")
        config.maxAge = 3600L

        val source = UrlBasedCorsConfigurationSource()
        source.registerCorsConfiguration("/**", config)
        return source
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return BCryptPasswordEncoder()
    }
}
